File: tests/sdk/nodejs/test/edge.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import {
  CreateBucketCommand,
  DeleteBucketCommand,
  DeleteObjectCommand,
  PutObjectCommand,
} from "@aws-sdk/client-s3";
import {
  newS3,
  requireReady,
  signedFetch,
  signedJsonTarget,
  uniquePrefix,
} from "../lib/helpers.mjs";

test("CloudFront CreateDistribution Deployed + optional edge GET", async (t) => {
  if (!(await requireReady(t))) return;
  const s3 = newS3();
  const prefix = uniquePrefix();
  const bucket = `${prefix}-cf-origin`.toLowerCase();
  const key = "docs/hi.txt";
  const payload = "edge-object-bytes";

  await s3.send(new CreateBucketCommand({ Bucket: bucket }));
  t.after(async () => {
    try {
      await s3.send(new DeleteObjectCommand({ Bucket: bucket, Key: key }));
    } catch {
      /* ignore */
    }
    try {
      await s3.send(new DeleteBucketCommand({ Bucket: bucket }));
    } catch {
      /* ignore */
    }
  });
  await s3.send(
    new PutObjectCommand({
      Bucket: bucket,
      Key: key,
      Body: Buffer.from(payload),
      ContentType: "text/plain",
    }),
  );

  const create = await signedJsonTarget(
    "cloudfront",
    "CloudFront_2016_01_28.CreateDistribution",
    {
      DistributionConfig: {
        CallerReference: `${prefix}-cf-ref`,
        Comment: "lab",
        Enabled: true,
        Origins: {
          Items: [{ Id: "o1", DomainName: bucket, OriginType: "s3" }],
        },
      },
    },
  );
  assert.equal(create.status, 200, create.body);
  const dist = create.json?.Distribution;
  assert.ok(dist?.Id, "missing Distribution.Id");
  assert.equal(dist.Status, "Deployed");
  assert.ok(
    dist.DomainName && String(dist.DomainName).includes("cloudfront"),
    `DomainName=${dist.DomainName}`,
  );

  const edge = await signedFetch(
    "cloudfront",
    "GET",
    `/cloudfront/${dist.Id}/${key}`,
  );
  const edgeBody = await edge.text();
  assert.equal(edge.status, 200, edgeBody);
  assert.equal(edgeBody, payload);
});

test("Transfer CreateServer ONLINE + CreateUser + lab PutFile/GetFile", async (t) => {
  if (!(await requireReady(t))) return;
  const prefix = uniquePrefix();
  const userName = "alice";

  const create = await signedJsonTarget(
    "transfer",
    "TransferService.CreateServer",
    { Protocols: ["SFTP"] },
  );
  assert.equal(create.status, 200, create.body);
  const serverId = create.json?.ServerId;
  assert.ok(serverId, "missing ServerId");
  t.after(async () => {
    try {
      await signedJsonTarget("transfer", "TransferService.DeleteUser", {
        ServerId: serverId,
        UserName: userName,
      });
    } catch {
      /* ignore */
    }
    try {
      await signedJsonTarget("transfer", "TransferService.DeleteServer", {
        ServerId: serverId,
      });
    } catch {
      /* ignore */
    }
  });

  const desc = await signedJsonTarget(
    "transfer",
    "TransferService.DescribeServer",
    { ServerId: serverId },
  );
  assert.equal(desc.status, 200, desc.body);
  const state = desc.json?.Server?.State ?? desc.json?.State;
  assert.equal(state, "ONLINE", desc.body);

  const user = await signedJsonTarget("transfer", "TransferService.CreateUser", {
    ServerId: serverId,
    UserName: userName,
  });
  assert.equal(user.status, 200, user.body);

  // Lab file API: TransferService.PutFile / GetFile (also available as
  // PUT/GET /transfer/{serverId}/home/{user}/path with SigV4 transfer).
  const relPath = `inbox/${prefix}.txt`;
  const put = await signedJsonTarget("transfer", "TransferService.PutFile", {
    ServerId: serverId,
    UserName: userName,
    Path: relPath,
    Body: "hello-transfer",
  });
  assert.equal(put.status, 200, put.body);

  const get = await signedJsonTarget("transfer", "TransferService.GetFile", {
    ServerId: serverId,
    UserName: userName,
    Path: relPath,
  });
  assert.equal(get.status, 200, get.body);
  const b64 = get.json?.BodyBase64;
  assert.ok(b64, "missing BodyBase64");
  assert.equal(Buffer.from(b64, "base64").toString("utf8"), "hello-transfer");

  const pathPut = await signedFetch(
    "transfer",
    "PUT",
    `/transfer/${serverId}/home/${userName}/path-style.txt`,
    "via-path",
    "application/octet-stream",
  );
  const pathPutBody = await pathPut.text();
  assert.equal(pathPut.status, 200, pathPutBody);
  const pathGet = await signedFetch(
    "transfer",
    "GET",
    `/transfer/${serverId}/home/${userName}/path-style.txt`,
  );
  const pathGetBody = await pathGet.text();
  assert.equal(pathGet.status, 200, pathGetBody);
  assert.equal(pathGetBody, "via-path");
});

test("Route53 Alias to CloudFront DomainName", async (t) => {
  if (!(await requireReady(t))) return;
  const s3 = newS3();
  const prefix = uniquePrefix();
  const bucket = `${prefix}-r53-cf`.toLowerCase();
  const zoneName = `${prefix.replace(/_/g, "-")}.example.com`;

  await s3.send(new CreateBucketCommand({ Bucket: bucket }));
  t.after(async () => {
    try {
      await s3.send(new DeleteBucketCommand({ Bucket: bucket }));
    } catch {
      /* ignore */
    }
  });

  const cf = await signedJsonTarget(
    "cloudfront",
    "CloudFront_2016_01_28.CreateDistribution",
    {
      DistributionConfig: {
        CallerReference: `${prefix}-r53-cf`,
        Comment: "lab",
        Enabled: true,
        Origins: {
          Items: [{ Id: "o1", DomainName: bucket, OriginType: "s3" }],
        },
      },
    },
  );
  assert.equal(cf.status, 200, cf.body);
  const domain = cf.json?.Distribution?.DomainName;
  assert.ok(domain, "missing CloudFront DomainName");

  const zone = await signedJsonTarget(
    "route53",
    "AWSRoute53.CreateHostedZone",
    {
      Name: zoneName,
      CallerReference: `${prefix}-hz`,
    },
  );
  assert.equal(zone.status, 200, zone.body);
  let zoneId = zone.json?.HostedZone?.Id || "";
  zoneId = zoneId.replace(/^\/hostedzone\//, "");
  assert.ok(zoneId, "missing HostedZone.Id");

  const change = await signedJsonTarget(
    "route53",
    "AWSRoute53.ChangeResourceRecordSets",
    {
      HostedZoneId: zoneId,
      ChangeBatch: {
        Changes: [
          {
            Action: "CREATE",
            ResourceRecordSet: {
              Name: `www.${zoneName}`,
              Type: "A",
              AliasTarget: {
                DNSName: domain,
                HostedZoneId: "Z2FDTNDATAQYW2",
                EvaluateTargetHealth: false,
              },
            },
          },
        ],
      },
    },
  );
  assert.equal(change.status, 200, change.body);

  const list = await signedJsonTarget(
    "route53",
    "AWSRoute53.ListResourceRecordSets",
    { HostedZoneId: zoneId },
  );
  assert.equal(list.status, 200, list.body);
  assert.match(list.body, /AliasTarget/i);
  assert.ok(
    list.body.toLowerCase().includes(String(domain).toLowerCase()),
    `list missing domain ${domain}: ${list.body}`,
  );
});

File: tests/sdk/go/edge_test.go
package sdk_test

import (
	"bytes"
	"context"
	"encoding/base64"
	"strings"
	"testing"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/s3"
)

func TestCloudFrontDeployedAndEdgeGET(t *testing.T) {
	requireReady(t)
	cfg := loadAWSConfig(t)
	s3c := newS3(t, cfg)
	ctx := context.Background()
	prefix := uniquePrefix(t)
	bucket := strings.ToLower(prefix + "-cf-origin")
	key := "docs/hi.txt"
	payload := []byte("edge-object-bytes")

	if _, err := s3c.CreateBucket(ctx, &s3.CreateBucketInput{Bucket: aws.String(bucket)}); err != nil {
		t.Fatalf("CreateBucket: %v", err)
	}
	t.Cleanup(func() {
		_, _ = s3c.DeleteObject(ctx, &s3.DeleteObjectInput{Bucket: aws.String(bucket), Key: aws.String(key)})
		_, _ = s3c.DeleteBucket(ctx, &s3.DeleteBucketInput{Bucket: aws.String(bucket)})
	})
	if _, err := s3c.PutObject(ctx, &s3.PutObjectInput{
		Bucket: aws.String(bucket), Key: aws.String(key), Body: bytes.NewReader(payload),
		ContentType: aws.String("text/plain"),
	}); err != nil {
		t.Fatalf("PutObject: %v", err)
	}

	status, body, parsed := signedJSONTarget(t, "cloudfront", "CloudFront_2016_01_28.CreateDistribution", map[string]any{
		"DistributionConfig": map[string]any{
			"CallerReference": prefix + "-cf-ref",
			"Comment":         "lab",
			"Enabled":         true,
			"Origins": map[string]any{
				"Items": []map[string]any{
					{"Id": "o1", "DomainName": bucket, "OriginType": "s3"},
				},
			},
		},
	})
	if status != 200 {
		t.Fatalf("CreateDistribution status=%d body=%s", status, body)
	}
	dist, _ := parsed["Distribution"].(map[string]any)
	id, _ := dist["Id"].(string)
	if id == "" {
		t.Fatalf("missing Id: %s", body)
	}
	if dist["Status"] != "Deployed" {
		t.Fatalf("Status=%v want Deployed body=%s", dist["Status"], body)
	}
	domain, _ := dist["DomainName"].(string)
	if domain == "" || !strings.Contains(domain, "cloudfront.noctaxris.local") {
		t.Fatalf("DomainName=%q body=%s", domain, body)
	}

	edgeStatus, edgeBody := signedHTTP(t, "cloudfront", "GET", "/cloudfront/"+id+"/"+key, nil, "")
	if edgeStatus != 200 {
		t.Fatalf("edge GET status=%d body=%s", edgeStatus, edgeBody)
	}
	if !bytes.Equal(edgeBody, payload) {
		t.Fatalf("edge body=%q want=%q", edgeBody, payload)
	}
}

func TestTransferLabPutGetFile(t *testing.T) {
	requireReady(t)
	prefix := uniquePrefix(t)

	status, body, parsed := signedJSONTarget(t, "transfer", "TransferService.CreateServer", map[string]any{
		"Protocols": []string{"SFTP"},
	})
	if status != 200 {
		t.Fatalf("CreateServer status=%d body=%s", status, body)
	}
	serverID, _ := parsed["ServerId"].(string)
	if serverID == "" {
		t.Fatalf("missing ServerId: %s", body)
	}
	t.Cleanup(func() {
		_, _, _ = signedJSONTarget(t, "transfer", "TransferService.DeleteUser", map[string]any{
			"ServerId": serverID, "UserName": "alice",
		})
		_, _, _ = signedJSONTarget(t, "transfer", "TransferService.DeleteServer", map[string]any{
			"ServerId": serverID,
		})
	})

	descStatus, descBody, descParsed := signedJSONTarget(t, "transfer", "TransferService.DescribeServer", map[string]any{
		"ServerId": serverID,
	})
	if descStatus != 200 {
		t.Fatalf("DescribeServer status=%d body=%s", descStatus, descBody)
	}
	state := ""
	if srv, ok := descParsed["Server"].(map[string]any); ok {
		state, _ = srv["State"].(string)
	}
	if state == "" {
		state, _ = descParsed["State"].(string)
	}
	if state != "ONLINE" {
		t.Fatalf("State=%q want ONLINE body=%s", state, descBody)
	}

	userStatus, userBody, _ := signedJSONTarget(t, "transfer", "TransferService.CreateUser", map[string]any{
		"ServerId": serverID, "UserName": "alice",
	})
	if userStatus != 200 {
		t.Fatalf("CreateUser status=%d body=%s", userStatus, userBody)
	}

	rel := "inbox/" + prefix + ".txt"
	putStatus, putBody, _ := signedJSONTarget(t, "transfer", "TransferService.PutFile", map[string]any{
		"ServerId": serverID, "UserName": "alice", "Path": rel, "Body": "hello-transfer",
	})
	if putStatus != 200 {
		t.Fatalf("PutFile status=%d body=%s", putStatus, putBody)
	}
	getStatus, getBody, getParsed := signedJSONTarget(t, "transfer", "TransferService.GetFile", map[string]any{
		"ServerId": serverID, "UserName": "alice", "Path": rel,
	})
	if getStatus != 200 {
		t.Fatalf("GetFile status=%d body=%s", getStatus, getBody)
	}
	b64, _ := getParsed["BodyBase64"].(string)
	raw, err := base64.StdEncoding.DecodeString(b64)
	if err != nil {
		t.Fatalf("BodyBase64: %v", err)
	}
	if string(raw) != "hello-transfer" {
		t.Fatalf("GetFile body=%q", raw)
	}
}

func TestRoute53AliasCloudFront(t *testing.T) {
	requireReady(t)
	cfg := loadAWSConfig(t)
	s3c := newS3(t, cfg)
	ctx := context.Background()
	prefix := uniquePrefix(t)
	bucket := strings.ToLower(prefix + "-r53-cf")
	zoneName := strings.ReplaceAll(prefix, "_", "-") + ".example.com"

	if _, err := s3c.CreateBucket(ctx, &s3.CreateBucketInput{Bucket: aws.String(bucket)}); err != nil {
		t.Fatalf("CreateBucket: %v", err)
	}
	t.Cleanup(func() {
		_, _ = s3c.DeleteBucket(ctx, &s3.DeleteBucketInput{Bucket: aws.String(bucket)})
	})

	cfStatus, cfBody, cfParsed := signedJSONTarget(t, "cloudfront", "CloudFront_2016_01_28.CreateDistribution", map[string]any{
		"DistributionConfig": map[string]any{
			"CallerReference": prefix + "-r53-cf",
			"Comment":         "lab",
			"Enabled":         true,
			"Origins": map[string]any{
				"Items": []map[string]any{
					{"Id": "o1", "DomainName": bucket, "OriginType": "s3"},
				},
			},
		},
	})
	if cfStatus != 200 {
		t.Fatalf("CreateDistribution status=%d body=%s", cfStatus, cfBody)
	}
	dist, _ := cfParsed["Distribution"].(map[string]any)
	domain, _ := dist["DomainName"].(string)
	if domain == "" {
		t.Fatalf("missing DomainName: %s", cfBody)
	}

	zoneStatus, zoneBody, zoneParsed := signedJSONTarget(t, "route53", "AWSRoute53.CreateHostedZone", map[string]any{
		"Name": zoneName, "CallerReference": prefix + "-hz",
	})
	if zoneStatus != 200 {
		t.Fatalf("CreateHostedZone status=%d body=%s", zoneStatus, zoneBody)
	}
	hz, _ := zoneParsed["HostedZone"].(map[string]any)
	zoneID, _ := hz["Id"].(string)
	zoneID = strings.TrimPrefix(zoneID, "/hostedzone/")
	if zoneID == "" {
		t.Fatalf("missing zone id: %s", zoneBody)
	}

	changeStatus, changeBody, _ := signedJSONTarget(t, "route53", "AWSRoute53.ChangeResourceRecordSets", map[string]any{
		"HostedZoneId": zoneID,
		"ChangeBatch": map[string]any{
			"Changes": []map[string]any{{
				"Action": "CREATE",
				"ResourceRecordSet": map[string]any{
					"Name": "www." + zoneName,
					"Type": "A",
					"AliasTarget": map[string]any{
						"DNSName": domain, "HostedZoneId": "Z2FDTNDATAQYW2", "EvaluateTargetHealth": false,
					},
				},
			}},
		},
	})
	if changeStatus != 200 {
		t.Fatalf("ChangeResourceRecordSets status=%d body=%s", changeStatus, changeBody)
	}

	listStatus, listBody, _ := signedJSONTarget(t, "route53", "AWSRoute53.ListResourceRecordSets", map[string]any{
		"HostedZoneId": zoneID,
	})
	if listStatus != 200 {
		t.Fatalf("ListResourceRecordSets status=%d body=%s", listStatus, listBody)
	}
	if !strings.Contains(string(listBody), "AliasTarget") ||
		!strings.Contains(strings.ToLower(string(listBody)), strings.ToLower(domain)) {
		t.Fatalf("list missing AliasTarget/domain: %s", listBody)
	}
}
